esp-config/src/bin/esp-config/main.rs

```diff
@@ -0,0 +1,310 @@
+        }
+        chip
+    };
+
+    // the "ESP_CONFIG_CHIP" hint env-var if present
+    let chip_from_config = || {
+        envs.get("ESP_CONFIG_CHIP")
+            .and_then(|chip_str| clap::ValueEnum::from_str(chip_str, true).ok())
+    };
+
+    // - if given as a parameter, use it
+    // - if there is a hint in the config.toml, use it
+    // - if we can infer it from metadata, use it
+    // otherwise, fail
+    let chip = chip_from_args
+        .or_else(|| chip_from_config())
+        .or_else(|| chip_from_meta());
+
+    if chip.is_none() {
+        return Err("No chip given or inferred. Try using the `--chip` argument.".into());
+    }
+
+    let mut configs = Vec::new();
+    let chip = esp_metadata::Config::for_chip(chip.as_ref().unwrap());
+    let features = vec![];
+    for krate in meta.packages {
+        let maybe_cfg = krate.manifest_path.parent().unwrap().join("esp_config.yml");
+        if maybe_cfg.exists() {
+            let yaml = std::fs::read_to_string(&maybe_cfg)?;
+            let (cfg, options) =
+                esp_config::evaluate_yaml_config(&yaml, Some(chip.clone()), features.clone(), true)
+                    .map_err(|e| {
+                        format!(
+                            "Error evaluating YAML config for crate {}: {}",
+                            krate.name, e
+                        )
+                    })?;
+
+            let crate_name = cfg.krate.clone();
+
+            let options: Vec<ConfigItem> = options
+                .iter()
+                .map(|cfg| {
+                    Ok::<ConfigItem, Box<dyn Error>>(ConfigItem {
+                        option: cfg.clone(),
+                        actual_value: {
+                            let key = cfg.full_env_var(&crate_name);
+                            let def_val = &cfg.default_value.to_string();
+                            let val = envs.get(&key).unwrap_or(def_val);
+
+                            let mut parsed_val = cfg.default_value.clone();
+                            parsed_val.parse_in_place(val).map_err(|_| {
+                                <std::string::String as Into<Box<dyn Error>>>::into(format!(
+                                    "Unable to parse '{val}' for option '{}'",
+                                    &cfg.name
+                                ))
+                            })?;
+                            parsed_val
+                        },
+                    })
+                })
+                .collect::<Result<Vec<ConfigItem>, Box<dyn Error>>>()?;
+
+            configs.push(CrateConfig {
+                name: crate_name.clone(),
+                checks: cfg.checks.clone(),
+                options,
+            });
+        }
+    }
+
+    if configs.is_empty() {
+        return Err("No config files found.".into());
+    }
+
+    Ok(configs)
+}
+
+fn check_after_changes(
+    work_dir: &PathBuf,
+    chip_from_args: Option<esp_metadata::Chip>,
+    config_file: Option<String>,
```

Comment:
Take `Option<&str>`. We shouldn't be this liberal with cloning/allocating memory for no good reason.